content/en/blog/2022/instrument-kafka-clients/index.md

```diff
@@ -0,0 +1,252 @@
+
+```java
+props.setProperty(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingProducerInterceptor.class.getName());
+```
+
+Use the `TracingConsumerInterceptor` for the consumer in order to create a "receive" span automatically, each time a message is received.
+
+```java
+props.setProperty(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+```
+
+### Wrapping the clients
+
+The other way is by wrapping the Kafka client with a tracing enabled one.
+
+On the producer side, assuming you have a `Producer<K, V>` instance, you can wrap it in the following way.
+
+```java
+KafkaTelemetry telemetry = KafkaTelemetry.create(GlobalOpenTelemetry.get());
+Producer<String, String> tracingProducer = telemetry.wrap(producer);
+```
+
+Then use the `tracingProducer` as usual for sending messages to the Kafka cluster.
+
+On the consumer side, assuming you have a `Consumer<K, V>` instance, you can wrap it in the following way.
+
+```java
+KafkaTelemetry telemetry = KafkaTelemetry.create(GlobalOpenTelemetry.get());
+Consumer<String, String> tracingConsumer = telemetry.wrap(this.consumer);
+```
+
+Then use the `tracingConsumer` as usual for receiving messages from the Kafka cluster.
+
+## Instrumenting by using the agent
+
+Another way is by adding tracing to your application with no changes or additions into your application code.
+You also don't need to add any dependencies to OpenTelemetry specific libraries.
+It is possible by using the OpenTelemetry agent you can download from [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases).
+This agent has to run alongside your application in order to inject the logic for tracing messages sent and received to/from a Kafka cluster.
+
+Run the producer application in the following way.
+
+```shell
+java -javaagent:path/to/opentelemetry-javaagent.jar \
+      -Dotel.service.name=my-kafka-service \
+      -Dotel.traces.exporter=jaeger \
+      -Dotel.metrics.exporter=none \
+      -jar kafka-producer-agent/target/kafka-producer-agent-1.0-SNAPSHOT-jar-with-dependencies.jar
+```
+
+Run the consumer application similarly.
+
+```shell
+java -javaagent:path/to/opentelemetry-javaagent.jar \
+      -Dotel.service.name=my-kafka-service \
+      -Dotel.traces.exporter=jaeger \
+      -Dotel.metrics.exporter=none \
+      -Dotel.instrumentation.messaging.experimental.receive-telemetry.enabled=true \
+      -jar kafka-consumer-agent/target/kafka-consumer-agent-1.0-SNAPSHOT-jar-with-dependencies.jar
+```
+
+The agent leverages the auto-configuration SDK extension, by setting the main parameters through system properties.
+
+## Instrumentation in action
+
+In order to practice with the instrumentation of Kafka clients by using the provided examples, first of all you need an Apache Kafka cluster.
+The easiest way is to download it from the [official website](https://kafka.apache.org/downloads) and running just one ZooKeeper node together with one Kafka broker.
```

Comment:
isn't there some docker images people can use, or even a docker compose that build everything for them?

Reply:
There are ... some of them use Kafka images from the Strimzi project, others use the Confluent one. I would like to be less opinionated and using directly the upstream project.

Reply:
What I wanted to say: is there one in your sample github repo or in one you create for this blog post, people can just copy that and run it (like we did it in the nginx blog where we give something people can follow step-by-step from 0)

Reply:
No, because most of my development with Kafka is always with running it locally without Docker compose (I don't have it installed at all :-)). The other is about using Kubernetes when I worked on Strimzi but using Kubernetes is really useless and too much for the purpose of the blog post.

Reply:
ok, then let's leave it like that!